private/buf/bufgraph/builder.go

```diff
@@ -0,0 +1,213 @@
+	node Node,
+	workspace bufmodule.Workspace,
+	graph *dag.Graph[Node],
+) ([]bufanalysis.FileAnnotation, error) {
+	image, fileAnnotations, err := b.imageBuilder.Build(
+		ctx,
+		module,
+		bufimagebuild.WithWorkspace(workspace),
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(fileAnnotations) > 0 {
+		return fileAnnotations, nil
+	}
+	for _, imageModuleDependency := range bufimage.ImageModuleDependencies(image) {
+		dependencyNode := newNodeForImageModuleDependency(imageModuleDependency)
+		if imageModuleDependency.IsDirect() {
+			graph.AddEdge(node, dependencyNode)
+		}
+		dependencyModule, err := b.getModuleForImageModuleDependency(
+			ctx,
+			imageModuleDependency,
+			workspace,
+		)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: deal with the case where there are differing commits for a given ModuleIdentity.
+		fileAnnotations, err := b.buildForModule(
+			ctx,
+			dependencyModule,
+			dependencyNode,
+			workspace,
+			graph,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if len(fileAnnotations) > 0 {
+			return fileAnnotations, nil
+		}
+	}
+	return nil, nil
+}
+
+func (b *builder) getModuleForImageModuleDependency(
+	ctx context.Context,
+	imageModuleDependency bufimage.ImageModuleDependency,
+	workspace bufmodule.Workspace,
+) (bufmodule.Module, error) {
+	moduleIdentity := imageModuleDependency.ModuleIdentity()
+	commit := imageModuleDependency.Commit()
+	if workspace != nil {
+		module, ok := workspace.GetModule(moduleIdentity)
+		if ok {
+			return module, nil
+		}
+	}
+	if commit == "" {
+		// TODO: can we error here? The only
+		// case we should likely not have a commit is when we are using a workspace.
+		// There's no enforcement of this property, so erroring here is a bit weird,
+		// but it might be better to check our assumptions and figure out if there
+		// are exceptions after the fact, as opposed to resolving a ModulePin for
+		// main when we don't know if main is what we want.
+		return nil, fmt.Errorf("had ModuleIdentity %v with no associated commit, but did not have the module in a workspace", moduleIdentity)
+	}
+	moduleReference, err := bufmoduleref.NewModuleReference(
+		moduleIdentity.Remote(),
+		moduleIdentity.Owner(),
+		moduleIdentity.Repository(),
+		commit,
+	)
+	if err != nil {
+		return nil, err
+	}
+	modulePin, err := b.moduleResolver.GetModulePin(
+		ctx,
+		moduleReference,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return b.moduleReader.GetModule(
+		ctx,
+		modulePin,
+	)
+}
+
+func newNodeForImageModuleDependency(imageModuleDependency bufimage.ImageModuleDependency) Node {
+	return Node{
+		Remote:     imageModuleDependency.ModuleIdentity().Remote(),
+		Owner:      imageModuleDependency.ModuleIdentity().Owner(),
+		Repository: imageModuleDependency.ModuleIdentity().Repository(),
+		Commit:     imageModuleDependency.Commit(),
+	}
+}
+
+func newNodeForModule(module bufmodule.Module, i int) Node {
+	// TODO: deal with unnamed Modules
+	node := Node{}
```

Comment:
```suggestion
	var node Node
```
This is just the zero value.

Reply:
Changed